### PR TITLE
feat(iOS): screensharing support

### PR DESCRIPTION
**Pull request by alexbumbu**

### CODE REVIEW

File: ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.m [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.m:3]: ditto

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleHandler.h:3]: 👀

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:3]: 👀

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    // User has requested to finish the broadcast.
+  [[DarwinNotificationCenter sharedInstance] postNotificationWithName:kBroadcastStoppedNotification];
+  [self.clientConnection close];
+}
+
+- (void)processSampleBuffer:(CMSampleBufferRef)sampleBuffer withType:(RPSampleBufferType)sampleBufferType {
+  static NSUInteger frameCount = 0;
+  switch (sampleBufferType) {
+    case RPSampleBufferTypeVideo:
+      if (++frameCount%3 == 0 && self.uploader.isReady) {
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:70]: Let's leave a comment here.

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    default:
+      break;
+  }
+}
+
+// MARK: Private Methods
+
+- (NSString *)socketFilePath {
+    NSURL *sharedContainer = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:@"group.org.jitsi.meet.appgroup"];
+    NSString *socketFilePath = [[sharedContainer URLByAppendingPathComponent:@"rtc_SSFD"] path];
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:84]: Let's leave a comment here mentioning this has to match with what RN WebRTC does.

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    default:
+      break;
+  }
+}
+
+// MARK: Private Methods
+
+- (NSString *)socketFilePath {
+    NSURL *sharedContainer = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:@"group.org.jitsi.meet.appgroup"];
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:83]: Can we fish it from the Plist?

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleUploader.h:3]: 👀

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m [hunk 8]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleUploader.m:3]: 👀

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+ * limitations under the License.
+ */
+
+#import <MessageUI/MessageUI.h>
+#import <ReplayKit/ReplayKit.h>
+
+#import "SampleUploader.h"
+#import "SocketConnection.h"
+
+static const NSInteger kBufferMaxLenght = 1000 * 1024;
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleUploader.m:24]: ❓

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    weakSelf.isReady = true;
+  };
+  self.connection.streamHasSpaceAvailable = ^{
+    dispatch_async(weakSelf.serialQueue, ^{
+      weakSelf.isReady = ![weakSelf sendData];
+    });
+  };
+}
+
+- (void)prepareSample:(CMSampleBufferRef)sampleBuffer {
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SampleUploader.m:80]: Document what this does please.

File: ios/app/JitsiMeetBroadcast Extension/SocketConnection.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SocketConnection.h:3]: 👀

File: ios/app/JitsiMeetBroadcast Extension/SocketConnection.m [hunk 12]
```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/app/JitsiMeetBroadcast Extension/SocketConnection.m:3]: 👀

File: ios/app/app.xcodeproj/project.pbxproj [hunk 13]
```diff
@@ -770,6 +880,70 @@
...
+				CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER = YES;
+				CLANG_WARN_UNGUARDED_AVAILABILITY = YES_AGGRESSIVE;
+				CODE_SIGN_ENTITLEMENTS = "JitsiMeetBroadcast Extension/JitsiMeetBroadcast Extension.entitlements";
+				CODE_SIGN_IDENTITY = "iPhone Developer";
+				CODE_SIGN_STYLE = Automatic;
+				DEBUG_INFORMATION_FORMAT = dwarf;
+				DEVELOPMENT_TEAM = FC967L3QRG;
+				GCC_C_LANGUAGE_STANDARD = gnu11;
+				INFOPLIST_FILE = "JitsiMeetBroadcast Extension/Info.plist";
+				IPHONEOS_DEPLOYMENT_TARGET = 14.2;
```
Comment (saghul) [inline ios/app/app.xcodeproj/project.pbxproj:900]: 11

File: ios/app/app.xcodeproj/project.pbxproj [hunk 14]
```diff
@@ -770,6 +880,70 @@
...
+				CLANG_WARN_UNGUARDED_AVAILABILITY = YES_AGGRESSIVE;
+				CODE_SIGN_ENTITLEMENTS = "JitsiMeetBroadcast Extension/JitsiMeetBroadcast Extension.entitlements";
+				CODE_SIGN_IDENTITY = "iPhone Developer";
+				CODE_SIGN_STYLE = Automatic;
+				COPY_PHASE_STRIP = NO;
+				DEBUG_INFORMATION_FORMAT = "dwarf-with-dsym";
+				DEVELOPMENT_TEAM = FC967L3QRG;
+				GCC_C_LANGUAGE_STANDARD = gnu11;
+				INFOPLIST_FILE = "JitsiMeetBroadcast Extension/Info.plist";
+				IPHONEOS_DEPLOYMENT_TARGET = 14.2;
```
Comment (saghul) [inline ios/app/app.xcodeproj/project.pbxproj:933]: 11

File: ios/app/src/AppDelegate.m [hunk 15]
```diff
@@ -35,6 +35,7 @@ -             (BOOL)application:(UIApplication *)application
 
     jitsiMeet.defaultConferenceOptions = [JitsiMeetConferenceOptions fromBuilder:^(JitsiMeetConferenceOptionsBuilder *builder) {
         [builder setFeatureFlag:@"resolution" withValue:@(360)];
+        [builder setFeatureFlag:@"ios.screensharing.extension" withValue:@"org.jitsi.meet.broadcast.extension"];
```
Comment (saghul) [inline ios/app/src/AppDelegate.m:38]: Let's have this in the plist as well please so the WebRTC part needs no parameters.

File: ios/sdk/src/ScheenshareEventEmiter.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/sdk/src/ScheenshareEventEmiter.h:3]: 👀

File: ios/sdk/src/ScheenshareEventEmiter.m [hunk 17]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
Comment (saghul) [inline ios/sdk/src/ScheenshareEventEmiter.m:3]: 👀

File: react/features/base/flags/constants.js [hunk 18]
```diff
@@ -1,5 +1,9 @@
 // @flow
 
+/**
+ */
+export const IOS_SCREENSHARING_EXTENSION = 'ios.screensharing.extension';
```
Comment (saghul) [inline react/features/base/flags/constants.js:5]: This can go.

File: react/features/toolbox/components/native/ScreenSharingIosButton.js [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * Whether video is currently muted or not.
+     */
+    _screensharing: boolean,
+
+    /**
+     * The preferredExtension extension for configuring ScreenCapturePickerView.
+     */
+    _extension: string,
```
Comment (saghul) [inline react/features/toolbox/components/native/ScreenSharingIosButton.js:28]: Move to plist

File: react/features/toolbox/components/native/ScreenSharingIosButton.js [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * @param {Object} state - The Redux state.
+ * @private
+ * @returns {{
+ *     _disabled: boolean,
+ * }}
+ */
+function _mapStateToProps(state): Object {
+    return {
+        _extension: getFeatureFlag(state, IOS_SCREENSHARING_EXTENSION, ''),
```
Comment (saghul) [inline react/features/toolbox/components/native/ScreenSharingIosButton.js:135]: Override `visible` to not show up in iOS < 12

**Review by saghul — COMMENTED** [review]
Some minor comments, PTAL!

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m (hunk 4 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:70]: done

File: ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.h [hunk 21]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
```
[the remark this answers is not in the document]
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.h:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.m (hunk 1 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/DarwinNotificationCenter.m:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.h (hunk 2 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.h:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m (hunk 3 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m (hunk 5 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.m:84]: done

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.h (hunk 7 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleUploader.h:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m (hunk 8 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleUploader.m:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m (hunk 10 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleUploader.m:80]: done. Also, I've slightly updated the logic so this function only does the image processing and wrapping into a CFHTTPMessageRef. Storing the data & reseting the state part is moved to `sendSample: `

File: ios/app/JitsiMeetBroadcast Extension/SocketConnection.h (hunk 11 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SocketConnection.h:3]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SocketConnection.m (hunk 12 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SocketConnection.m:3]: fixed

File: ios/app/app.xcodeproj/project.pbxproj (hunk 13 above)
Comment (alexbumbu) [reply ios/app/app.xcodeproj/project.pbxproj:900]: fixed

File: ios/app/app.xcodeproj/project.pbxproj (hunk 14 above)
Comment (alexbumbu) [reply ios/app/app.xcodeproj/project.pbxproj:933]: fixed

File: ios/app/src/AppDelegate.m (hunk 15 above)
Comment (alexbumbu) [reply ios/app/src/AppDelegate.m:38]: moved to WebRTC

File: ios/sdk/src/ScheenshareEventEmiter.h (hunk 16 above)
Comment (alexbumbu) [reply ios/sdk/src/ScheenshareEventEmiter.h:3]: fixed

File: ios/sdk/src/ScheenshareEventEmiter.m (hunk 17 above)
Comment (alexbumbu) [reply ios/sdk/src/ScheenshareEventEmiter.m:3]: fixed

File: react/features/base/flags/constants.js (hunk 18 above)
Comment (alexbumbu) [reply react/features/base/flags/constants.js:5]: removed

File: react/features/toolbox/components/native/ScreenSharingIosButton.js (hunk 19 above)
Comment (alexbumbu) [reply react/features/toolbox/components/native/ScreenSharingIosButton.js:28]: done

File: react/features/toolbox/components/native/ScreenSharingIosButton.js (hunk 20 above)
Comment (alexbumbu) [reply react/features/toolbox/components/native/ScreenSharingIosButton.js:135]: done

File: ios/app/JitsiMeetBroadcast Extension/SampleUploader.m (hunk 9 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleUploader.m]: fixed

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m (hunk 6 above)
Comment (alexbumbu) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.m]: no quick way for doing this. We could try looking at the project folder structure and read it that way, but it's not safe for future versions.

**Review by saghul — APPROVED** [review]

File: ios/app/JitsiMeetBroadcast Extension/SampleHandler.m (hunk 6 above)
Comment (saghul) [reply ios/app/JitsiMeetBroadcast Extension/SampleHandler.m]: Ack.

**Review by saghul — APPROVED** [review]

**Comment by saghul** [comment]
Jenkins add to whitelist.

**Comment by saghul** [comment]
Jenkins please test this please.

**Pull request merged**